### PR TITLE
Implement exporting megolm sessions to a file

### PR DESCRIPTION
**Pull request by TobiasFella**

### CODE REVIEW

**Comment by TobiasFella** [comment]
test added, should be ready now

File: Quotient/keyimport.cpp [hunk 1]
```diff
@@ -87,8 +87,71 @@ KeyImport::Error KeyImport::importKeys(QString data, const QString& passphrase,
...
     return Success;
 }
 
+Quotient::Expected<QByteArray, KeyImport::Error> KeyImport::encrypt(QJsonArray sessions, const QString& passphrase)
+{
+    auto plainText = QJsonDocument(sessions).toJson(QJsonDocument::Compact);
+
+    auto salt = getRandom(AesBlockSize);
+    auto iv = getRandom(AesBlockSize);
+    quint32 rounds = 200000; // spec: "N should be at least 100,000";
```
Comment (KitsuneRal) [inline Quotient/keyimport.cpp]:
```suggestion
    quint32 rounds = 200'000; // spec: "N should be at least 100,000";
```

File: Quotient/database.cpp [hunk 2]
```diff
@@ -238,6 +239,41 @@ void Database::migrateTo9()
...
+    qCDebug(DATABASE) << "Migrating database to version 10";
+
+    transaction();
+
+    execute(QStringLiteral("ALTER TABLE inbound_megolm_sessions ADD senderClaimedEd25519Key TEXT;"));
+
+    auto query = prepareQuery("SELECT senderKey FROM inbound_megolm_sessions;"_ls);
+    execute(query);
+
+    QSet<QString> keys;
```
Comment (KitsuneRal) [inline Quotient/database.cpp]: Do you actually need to go through keys in order? Wouldn't `QStringList` suffice?

File: Quotient/database.cpp [hunk 3]
```diff
@@ -238,6 +239,41 @@ void Database::migrateTo9()
...
+        keys += query.value("senderKey"_ls).toString();
+    }
+    for (const auto& key : keys) {
+        auto query = prepareQuery("SELECT edKey FROM tracked_devices WHERE curveKey=:curveKey;"_ls);
+        query.bindValue(":curveKey"_ls, key);
+        execute(query);
+        if (!query.next()) {
+            continue;
+        }
+        const auto &edKey = query.value("edKey"_ls).toByteArray();
```
Comment (KitsuneRal) [inline Quotient/database.cpp]: Avoiding variables shadowing each other. Eventually, I hope we'll get a generic method to retrieve a single value from the database - we've already got a lot of boilerplate code doing that. Not necessarily for this PR though.
```suggestion
        auto edKeyQuery = prepareQuery("SELECT edKey FROM tracked_devices WHERE curveKey=:curveKey;"_ls);
        edKeyquery.bindValue(":curveKey"_ls, key);
        execute(edKeyQuery);
        if (!edKeyQuery.next()) {
            continue;
        }
        const auto &edKey = edKeyQuery.value("edKey"_ls).toByteArray();
```

File: Quotient/keyimport.h [hunk 4]
```diff
@@ -35,10 +35,12 @@ class QUOTIENT_API KeyImport : public QObject
 
     Q_INVOKABLE Error importKeys(QString data, const QString& passphrase,
                                  const Quotient::Connection* connection);
+    Q_INVOKABLE Quotient::Expected<QByteArray, Error> exportKeys(const QString& passphrase, Quotient::Connection* connection);
```
Comment (KitsuneRal) [inline Quotient/keyimport.h]: I think constifying won't interfere with `Q_INVOKABLE`, will it?
```suggestion
    Q_INVOKABLE Quotient::Expected<QByteArray, Error> exportKeys(const QString& passphrase, const Quotient::Connection* connection);
```

File: Quotient/keyimport.cpp [hunk 5]
```diff
@@ -87,8 +87,71 @@ KeyImport::Error KeyImport::importKeys(QString data, const QString& passphrase,
...
+        base64.insert(i, "\n");
+        i++;
+    }
+    base64.prepend("-----BEGIN MEGOLM SESSION DATA-----\n"_ls);
+    base64.append("\n-----END MEGOLM SESSION DATA-----\n"_ls);
+    return base64;
+}
+
+
+Quotient::Expected<QByteArray, KeyImport::Error> KeyImport::exportKeys(const QString& passphrase, Connection* connection)
```
Comment (KitsuneRal) [inline Quotient/keyimport.cpp]:
```suggestion
Quotient::Expected<QByteArray, KeyImport::Error> KeyImport::exportKeys(const QString& passphrase, const Connection* connection)
```

File: Quotient/keyimport.cpp [hunk 6]
```diff
@@ -87,8 +87,71 @@ KeyImport::Error KeyImport::importKeys(QString data, const QString& passphrase,
...
+    QByteArray base64 = data.toBase64();
+
+    // Limit base64 line length. 96 is chosen arbitrarily.
+    for (auto i = 96; i < base64.length(); i += 96) {
+        base64.insert(i, "\n");
+        i++;
+    }
+    base64.prepend("-----BEGIN MEGOLM SESSION DATA-----\n"_ls);
+    base64.append("\n-----END MEGOLM SESSION DATA-----\n"_ls);
+    return base64;
```
Comment (KitsuneRal) [inline Quotient/keyimport.cpp]: Is
```suggestion
       return "-----BEGIN MEGOLM SESSION DATA-----\n"_ls
           % (std::views::all(data.toBase64()) | std::views::chunk(96)
              | std::views::join_with('\n') | std::ranges::to<QByteArray>())
           % "\n-----END MEGOLM SESSION DATA-----\n"_ls;

```
too scary?

**Review by KitsuneRal — COMMENTED** [review]
A few cleanup things to begin with.

File: Quotient/database.cpp (hunk 2 above)
Comment (TobiasFella) [reply Quotient/database.cpp]: this is (supposed to be) an optimization over using a QStringList, considering that there are (probably) on average >> 1 session per sender key in there, and we need each sender key only once.

File: Quotient/keyimport.cpp (hunk 6 above)
Comment (TobiasFella) [reply Quotient/keyimport.cpp]: nope, i love it. also, thanks for telling me about std::views::chunk :)

Comment (KitsuneRal) [reply Quotient/keyimport.cpp]: You're most welcome :)

Comment (KitsuneRal) [reply Quotient/keyimport.cpp]: Oh wait, it's actually even better than that I think:
```suggestion
       return "-----BEGIN MEGOLM SESSION DATA-----\n"_ls
           % (std::views::chunk(data.toBase64(), 96) | std::views::join_with('\n')
              | std::ranges::to<QByteArray>())
           % "\n-----END MEGOLM SESSION DATA-----\n"_ls;
```

File: Quotient/database.cpp (hunk 2 above)
Comment (KitsuneRal) [reply Quotient/database.cpp]: If your concern is uniqueness then `QHash` would be faster. But in this particular case I would simply use [`SELECT DISTINCT`](https://www.sqlitetutorial.net/sqlite-distinct/) on the database side then.

File: Quotient/keyimport.cpp [hunk 7]
```diff
@@ -87,8 +87,71 @@ KeyImport::Error KeyImport::importKeys(QString data, const QString& passphrase,
...
     }
     return Success;
 }
 
+Quotient::Expected<QByteArray, KeyImport::Error> KeyImport::encrypt(QJsonArray sessions, const QString& passphrase)
+{
+    auto plainText = QJsonDocument(sessions).toJson(QJsonDocument::Compact);
+
+    auto salt = getRandom(AesBlockSize);
+    auto iv = getRandom(AesBlockSize);
```
Comment (KitsuneRal) [inline Quotient/keyimport.cpp]:
```suggestion
    auto salt = getRandom(AesBlockSize);
    auto iv = getRandom(AesBlockSize);
```
```suggestion
    auto salt = getRandom<AesBlockSize>();
    auto iv = getRandom<AesBlockSize>();
```

File: Quotient/keyimport.cpp (hunk 6 above)
Comment (KitsuneRal) [reply Quotient/keyimport.cpp]: Grr, I keep forgetting that `std::ranges::to` is only available in very new stdlibs - libstdc++ 14, in particular. Though Xcode 15 is supposed to already have it; not sure why CI still fails there.

File: Quotient/keyimport.cpp [hunk 8]
```diff
@@ -87,8 +90,65 @@ KeyImport::Error KeyImport::importKeys(QString data, const QString& passphrase,
...
+    if (!mac.has_value()) {
+        qCWarning(E2EE) << "Failed to calculate MAC" << mac.error();
+        return OtherError;
+    }
+    data.append(mac.value());
+
+       return "-----BEGIN MEGOLM SESSION DATA-----\n"_ba
+           % (std::views::chunk(data.toBase64(), 96) | std::views::join_with('\n')
+              | std::ranges::to<QByteArray>())
+           % "\n-----END MEGOLM SESSION DATA-----\n"_ba;
```
Comment (KitsuneRal) [inline Quotient/keyimport.cpp]: Ok, here's a workaround while we wait for `std::ranges::to`:
```suggestion
    // TODO: use std::ranges::to() once it's available from all stdlibs Quotient builds with
   return "-----BEGIN MEGOLM SESSION DATA-----\n"_ba
          % std::ranges::fold_left(std::views::chunk(data.toBase64(), 96)
                                       | std::views::join_with('\n'),
                                   QByteArray{}, std::plus<>())
          % "\n-----END MEGOLM SESSION DATA-----\n"_ba;
```

File: Quotient/keyimport.cpp [hunk 9]
```diff
@@ -87,8 +90,65 @@ KeyImport::Error KeyImport::importKeys(QString data, const QString& passphrase,
...
+}
+
+
+Quotient::Expected<QByteArray, KeyImport::Error> KeyImport::exportKeys(const QString& passphrase, const Connection* connection)
+{
+    QJsonArray sessions;
+    for (const auto& room : connection->allRooms()) {
+        for (const auto &session : room->exportMegolmSessions()) {
+            return {};
```
Comment (KitsuneRal) [inline Quotient/keyimport.cpp]: ???

File: Quotient/room.cpp [hunk 10]
```diff
@@ -3444,3 +3447,36 @@ void Room::startVerification()
...
+            {"forwarding_curve25519_key_chain"_ls, QJsonArray()},
+            {"room_id"_ls, id()},
+            {"sender_claimed_keys"_ls, QJsonObject{
+                {"ed25519"_ls, connection()->database()->edKeyForMegolmSession(QString::fromLatin1(value.sessionId()))},
+            }},
+            {"sender_key"_ls, connection()->database()->senderKeyForMegolmSession(QString::fromLatin1(value.sessionId()))},
+            {"session_id"_ls, QString::fromLatin1(value.sessionId())},
+            {"session_key"_ls, QString::fromLatin1(session.value())},
+        };
+        if (json["sender_claimed_keys"_ls]["ed25519"_ls].toString().isEmpty() || json["sender_key"_ls].toString().isEmpty()) {
```
Comment (KitsuneRal) [inline Quotient/room.cpp]:
```suggestion
        const auto senderClaimedKey = connection()->database()->edKeyForMegolmSession(QString::fromLatin1(value.sessionId()));
        const auto senderKey = connection()->database()->senderKeyForMegolmSession(QString::fromLatin1(value.sessionId()));
        const auto json = QJsonObject {
            {"algorithm"_ls, "m.megolm.v1.aes-sha2"_ls},
            {"forwarding_curve25519_key_chain"_ls, QJsonArray()},
            {"room_id"_ls, id()},
            {"sender_claimed_keys"_ls, QJsonObject{ {"ed25519"_ls, senderClaimedKey} }},
            {"sender_key"_ls, senderKey},
            {"session_id"_ls, QString::fromLatin1(value.sessionId())},
            {"session_key"_ls, QString::fromLatin1(session.value())},
        };
        if (senderClaimedKey.isEmpty() || senderKey.isEmpty()) {
```

File: Quotient/keyimport.cpp (hunk 9 above)
Comment (TobiasFella) [reply Quotient/keyimport.cpp]: leftover from testing :facepalm:

File: Quotient/room.cpp [hunk 11]
```diff
@@ -3444,3 +3447,36 @@ void Room::startVerification()
     d->pendingKeyVerificationSession = new KeyVerificationSession(this);
     emit d->connection->newKeyVerificationSession(d->pendingKeyVerificationSession);
 }
+
+QJsonArray Room::exportMegolmSessions()
+{
+    QJsonArray sessions;
+    for (auto &[key, value] : d->groupSessions) {
```
Comment (KitsuneRal) [inline Quotient/room.cpp]:
```suggestion
    for (const auto& [key, value] : d->groupSessions) {
```
just to make sure :)

**Comment by KitsuneRal** [comment]
Aside from things above, LGTM.

File: Quotient/keyimport.cpp [hunk 12]
```diff
@@ -87,8 +90,66 @@ KeyImport::Error KeyImport::importKeys(QString data, const QString& passphrase,
...
+    auto mac = hmacSha256(key_view_t(keys.value().begin() + 32, 32), data);
+    if (!mac.has_value()) {
+        qCWarning(E2EE) << "Failed to calculate MAC" << mac.error();
+        return OtherError;
+    }
+    data.append(mac.value());
+
+    // TODO: use std::ranges::to() once it's available from all stdlibs Quotient builds with
+   return "-----BEGIN MEGOLM SESSION DATA-----\n"_ba
+          % std::ranges::fold_left(std::views::chunk(data.toBase64(), 96)
```
Comment (KitsuneRal) [inline Quotient/keyimport.cpp]: I just discovered that the macOS stdlib is quite behind on ranges - it has no `fold_left` and no `views::chunk`. I'll look into another workaround but this already seems to take more time and effort than it's worth. If I don't have anything good soon, let's just revert to your original code and revisit it in the version after 0.9.

**Review by KitsuneRal — APPROVED** [review]

**Pull request merged**